### ISSUE TITLE
win32 wheel package suitable for Azure Functions

### ISSUE THREAD

**Issue by flamby** [Azure/azure-sdk-for-python#1044]
Hi,

Currently, I was not able to install the sdk in Azure Functions.
It seems some dependencies like cffi needs Visual C++.
Error message :
```
  File "D:\home\site\wwwroot\QueueTriggerPython2\env\lib\site-packages\setuptools\msvc9_support.py", line 61, in query_vcvarsall

    message + ' Get it from http://aka.ms/vcpython27'

distutils.errors.DistutilsPlatformError: Microsoft Visual C++ 9.0 is required (Unable to find vcvarsall.bat). Get it from http://aka.ms/vcpython27
```
Could you provide win32 wheel packages suitable for Azure Functions?
i.e. azure sdk and its dependencies wheel packages

TIA

**Comment by lmazuel** [comment]
Hi @flamby 
When you say "install the SDK", what package do you need exactly?
Thanks,

**Comment by flamby** [comment]
Hi @lmazuel 
I would say all of them. Those installed by 
pip install azure --pre

**Comment by lmazuel** [comment]
I don't deny the issue, you're not the first and we are working on it. I was just thinking about help you asap versus help you in the long term. Because `azure` is a meta-package that does not contain code itself. By choosing carefully what version of real packages you install, you might be able to workaround the issue right now.
So short term, what do need exactly? Compute? Storage? etc.

Be sure I won't close the issue before I'm sure this works out of the box for everybody, but this does not imply that we won't try to make it work right now :)

**Comment by lmazuel** [comment]
Related issue: https://github.com/Azure/azure-storage-python/issues/219

**Comment by flamby** [comment]
@lmazuel I see. Thanks. 
I would say those 3 by order of priority (don't know their dependencies)
azure-mgmt-compute
azure-mgmt-network
azure-mgmt-resource

but i'm afraid that the problem can be in the external libs like cffi, will you (the project) provide them as well?

thanks anyway

**Comment by flamby** [comment]
I see a workaround for cffi in your related issue. And lots of other hints.
I'll share here what works for me.

**Comment by flamby** [comment]
The workaround works like a charm.
1. mkdir wheelhouse
2. [download cffi](https://pypi.python.org/pypi/cffi) official whl, rename it to cffi-1.10.0-cp27-none-win32.whl and upload it to the directory

- if default Azure Function python runtime, direct download : https://pypi.python.org/packages/8a/11/2c0ac16e44a097a2dcf04d7e1f91ec1c5fe51cf20b0c9e0ca1b4850e66e0/cffi-1.10.0-cp27-cp27m-win32.whl#md5=dce501481feb73eeafe880ef3a9dd96a

3. [download cryptography](https://pypi.python.org/pypi/cryptography) official whl, rename it to cryptography-1.8.1-cp27-none-win32.whl and upload it to the directory

- if default Azure Function python runtime, direct download : https://pypi.python.org/packages/9f/6b/f40266af48a9c57af01fb41b0c8d6076fc0cd75279053e86258ae33f0406/cryptography-1.8.1-cp27-cp27m-win32.whl#md5=e1f04c4482a668b5181202a9c65c90a0

4. requirements.txt
```
--find-links wheelhouse
cffi==1.10.0
cryptography==1.8.1
azure==2.0.0rc6
```
5. pip install -r requirements.txt
**That's it!**

So most wheel package installation issues come from this ABI tag in the wheel package name.

Now I'm facing an issue in the installation of msrestazure package, which seems a dependency for the following azure login method 
```python
from azure.common.credentials import ServicePrincipalCredentials
```
which requires keyring...

**Comment by lmazuel** [comment]
Thanks for all the details @flamby !
`msrestazure` is more than login, it's the entire RestAPI connection layer. What kind of issue do you have with keyring?

**Comment by lmazuel** [comment]
I just tested and have no issue installing keyring, could you share the error you have?

**Comment by silencev** [comment]
@flamby
I'm not sure if you are install packages in build-in Python of Azure Function App.
but I did not experience this issue in Function App.

1. I installed Python 3.5.2 via Kudu site extensions.
2. Navigate to directory D:\home\Python35
3. using Kudu cmd line to install packages:
```
python.exe -m pip install -r requirements.txt
```

here is my package list in requirements.txt:
```
setuptools==34.0.1
azure==2.0.0rc6
azure-keyvault==0.1.0
azure-storage==0.33.0
azure-mgmt-dns==1.0.0
azure-mgmt-servicebus==0.1.0
azure-mgmt-web==0.31.0
applicationinsights==0.10.0
msrest==0.4.4
msrestazure==0.4.7
pymongo==3.4.0
```

**Comment by flamby** [comment]
@silencev I'm using python 2.7 (default one).

I guess all this pip / wheel mess is fixed in python 3.x

@lmazuel did you test w/ python 2.7?
keyring pip installation went well, but loading module did not. i suspect it's due to the module installed as an egg file, instead of a directory. probably fixed also in 3.x, but I need 2.7 for some modules.

**Comment by lmazuel** [comment]
@silencev in Azure Functions you can't override Python yet

Your packages are tested in Py2.7, but maybe pip is old enough so keyring is not installed from the wheel. could you try to update pip and see what happen? I will check too now I know it's runtime and not installation.

**Comment by lmazuel** [comment]
@flamby 

I make it work:
![image](https://cloud.githubusercontent.com/assets/1050156/24217953/c1720508-0efe-11e7-8eb6-5ed7975e5aa1.png)

So, I followed your instructions, but you need first to update pip to latest one (I have 9.0.1 installed in my venv now). This will take care of keyring and latest azure packages. This might be enough as well to avoid your "none renaming" issue, need to try it.

Also in this test I installed directly `azure-mgmt-resource` and I used ADAL, but I have no reason to believe you can't use [any authentication you want](http://azure-sdk-for-python.readthedocs.io/en/latest/quickstart_authentication.html).

FYI @zooba @brettcannon

**Comment by lmazuel** [comment]
@flamby 

Actually, I just tested and pip upgrade is enough. So:
- Create a virtualenv and load it
- Update pip `pip install -U pip`
- Install the SDK as usual

No rename or manual download :)

**Comment by flamby** [comment]
@lmazuel good to know! should be documented then ;-)
as I'm running out of credits, i won't be able to test it yet.

**Comment by flamby** [comment]
@lmazuel indeed. I was finally able to test it and it works like a charm. thanks!

**Comment by 4c74356b41** [comment]
@lmazuel this doesnt really work terribly reliably, right? this wont scale or if the host changes this will need manual intervention?

**Comment by lmazuel** [comment]
A v2 of Azure Functions is coming for Python:
https://docs.microsoft.com/en-us/azure/azure-functions/supported-languages#languages-in-runtime-1x-and-2x

For the first examples I saw in Python v2, this kind of trouble will disapear.

**Comment by 4c74356b41** [comment]
@lmazuel do you know the timeline? also, that article doesnt list python as available for 2.0...

**Comment by 4c74356b41** [comment]
well, the article you linked doesnt support your claim in any way (i'm not saying you are wrong, i'm just saying that I dont see any evidence in that article that supports your claim). thanks @lmazuel
